lib/PuppeteerSharp/ExecutionContext.cs

```diff
@@ -22,44 +24,86 @@ public ExecutionContext(Session client, ContextPayload contextPayload, Func<dyna
         public string FrameId { get; internal set; }
         public bool IsDefault { get; internal set; }
 
-        public async Task<dynamic> Evaluate(string pageFunction, params object[] args)
+        public async Task<dynamic> EvaluateAsync(string script, params object[] args)
         {
-            var handle = await EvaluateHandleAsync(pageFunction, args);
+            var handle = await EvaluateHandleAsync(script, args);
             dynamic result = await handle.JsonValue();
             await handle.Dispose();
             return result;
         }
 
-        internal async Task<JSHandle> EvaluateHandleAsync(Func<object> pageFunction, object[] args)
+        public async Task<T> EvaluateAsync<T>(string script, params object[] args)
         {
-            throw new NotImplementedException();
+            var result = await EvaluateAsync(script, args);
+            return ((JToken)result).ToObject<T>();
         }
 
-        internal async Task<JSHandle> EvaluateHandleAsync(string pageFunction, object[] args)
+        internal async Task<JSHandle> EvaluateHandleAsync(string script, object[] args)
         {
-            if (!string.IsNullOrEmpty(pageFunction))
+            if (string.IsNullOrEmpty(script))
             {
-                dynamic remoteObject;
+                return null;
+            }
+
+            if (IsFunction(script))
+            {
+                return await EvaluateFunctionAsync(script, args);
+            }
+
+            return await EvaluateExpressionAsync(script);
+
+        }
 
-                try
+        private async Task<JSHandle> EvaluateExpressionAsync(string script)
+        {
+            dynamic remoteObject;
+
+            try
+            {
+                remoteObject = await _client.SendAsync("Runtime.evaluate", new Dictionary<string, object>()
                 {
-                    remoteObject = await _client.SendAsync("Runtime.evaluate", new Dictionary<string, object>()
-                    {
-                        ["expression"] = pageFunction,
-                        ["contextId"] = _contextId,
-                        ["returnByValue"] = false,
-                        ["awaitPromise"] = true
-                    });
-
-                    return ObjectHandleFactory(remoteObject.result);
-                }
-                catch (Exception ex)
+                    {"contextId", _contextId},
```

Comment:
I see you don't like the C# 6 dictionary initializer syntax ;)

Reply:
hahah I'm not used to that. On the other hand and don't have to have some files with one style and other files with another style. But yeah, we can refactor that in the future.